src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java

```diff
@@ -910,7 +910,7 @@ private Set<PatternDescription> reduceBindingPatterns(Type selectorType, Set<Pat
                     Set<PatternDescription> toAdd = new HashSet<>();
 
                     for (Type sup : types.directSupertypes(bpOne.type)) {
-                        ClassSymbol clazz = (ClassSymbol) sup.tsym;
+                        ClassSymbol clazz = (ClassSymbol) types.erasure(sup).tsym;
```

Comment:
general comment: I think that erasure should happen in the desugaring phase, we are already doing some erasure in Flow and all of it is related to patterns. I think that we should consider moving all these erasure to either TransTypes or TransPatterns

Reply:
While I agree that erased type attributes int the AST should happen after `Flow`, I don't think that's what is happening here. The erased types here are used by the exhaustiveness machinery, which is specified in terms of erased type. The erased type produce here, however, shouldn't be recorded in the AST.

But, and the same time, I am not sure if the erasure is correct here. It `sup` is a type variable with multiple bounds, we'll get only the first bound. I think we need to check if that make a difference at this place, I was trying to look, but didn't find any (yet).

Reply:
sure I agree my comment about erasure is a meta-comment, it would be desirable to do all the erasure after flow, unless it is not possible of course

Reply:
Added a test with multiple bounds IIUC @lahodaj: https://github.com/openjdk/jdk/pull/21606/commits/f9f9888674577222ae268fd14fd1fdf978d6418e

Still ok, it seems. Hm.